File: Sources/UI/FrezIntroSheet.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import SwiftUI

/// The note Frez asks every app on its Dyno API to show ONCE, the first time someone
/// picks the Dyno. The words are Donghyun Kim's, verbatim, in English in every locale;
/// only the title, footnote and button are ours.
///
/// Laid out as its author asked (2026-09-18, "a few line breaks to improve
/// readability"): every line its own paragraph with a visible gap (a break that only
/// shows when the line wraps is no break), the salutation as heading and the signature
/// closing it, so it reads as a letter.
///
/// Next is the only way out: swiping away would leave the selection half-made, and the
/// note is the maker's one ask. After Next the flag persists and it never reappears.
struct FrezIntroSheet: View {
    var onNext: () -> Void

    private let greeting = String(localized: "Hello,")
    private let introduction = String(localized: "I’m Donghyun Kim, founder of Frez.")

    /// The body, one paragraph per line he wrote.
    private let paragraphs: [String] = [
        String(localized: "We keep our hardware margins low because we believe everyone should have access to their own data."),
        String(localized: "Frez Pro helps us develop new features and provide reliable devices."),
        String(localized: "In fact, Frez Dyno was made possible by our early Pro subscribers."),
        String(localized: "If you enjoy using Frez Dyno and would like to support the future of Frez, please consider trying Frez Pro."),
        String(localized: "I hope Frez will be with you for many years to come."),
    ]

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: 10) {
                    MaterialCard(verticalPadding: 22) {
                        VStack(alignment: .leading, spacing: 18) {
                            VStack(alignment: .leading, spacing: 6) {
                                Text(greeting)
                                    .font(.title2.weight(.semibold))
                                Text(introduction)
                            }
                            ForEach(paragraphs, id: \.self) { paragraph in
                                Text(paragraph)
                            }
                            VStack(alignment: .leading, spacing: 4) {
                                Text("Thank you,")
                                Text("Donghyun Kim")
                                    .fontWeight(.semibold)
                            }
                            .padding(.top, 4)
                        }
                        .font(.body)
                        .lineSpacing(3)
                        .foregroundStyle(Ink.primary)
                        .fixedSize(horizontal: false, vertical: true)
                        .frame(maxWidth: .infinity, alignment: .leading)
                    }
                    Text("Shown once, the first time you pick the Dyno.")
                        .font(.footnote)
                        .foregroundStyle(Ink.tertiary)
                        .padding(.horizontal, 4)
                }
                .padding(.horizontal, Metrics.hPadding)
                .padding(.top, 4)
                .frame(maxWidth: Metrics.maxContentWidth)
                .frame(maxWidth: .infinity)
            }
            .background { AppBackground() }
            .safeAreaInset(edge: .bottom) {
                PrimaryGlassButton(title: String(localized: "Next"), action: onNext)
                    .padding(.horizontal, Metrics.hPadding)
                    .padding(.bottom, 4)
                    .frame(maxWidth: Metrics.maxContentWidth)
                    .frame(maxWidth: .infinity)
                    .accessibilityIdentifier("frez.intro.next")
            }
            .navigationTitle("A note from Frez")
            .navigationBarTitleDisplayMode(.inline)
        }
        .interactiveDismissDisabled(true)
        .presentationDetents([.large])
    }
}

/// `.sheet(item:)` wants an identity; the raw value is the only one a kind needs.
extension GaugeKind: Identifiable {
    var id: String { rawValue }
}

File: android/app/src/main/kotlin/run/nuri/getagrip/ui/settings/FrezIntroSheet.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.ui.settings

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.safeDrawingPadding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import run.nuri.getagrip.ui.components.PrimaryButton
import run.nuri.getagrip.ui.l10n.tr
import run.nuri.getagrip.ui.theme.InstrumentSurface
import run.nuri.getagrip.ui.theme.LocalGripPalette
import run.nuri.getagrip.ui.theme.Metrics
import run.nuri.getagrip.ui.theme.readablePageWidth

/// The note Frez asks every app on its Dyno API to show ONCE, the first time someone
/// picks the Dyno. A letter, so it reads as one: the words are Donghyun Kim's, verbatim,
/// and stay in English in every locale — only the title and the button are ours.
///
/// Next is the only way out. Swiping or backing out would leave the selection half-made,
/// and the note is the maker's one ask for lending its calibration service; after Next the
/// flag is persisted and the sheet never appears again on this device.
///
/// TRANSLATION NOTE: iOS presents this as a `.sheet` with `interactiveDismissDisabled` and
/// the `.large` detent. The Android twin of "a full-height presentation you cannot swipe
/// away" is a `Dialog` with `usePlatformDefaultWidth = false` — the same shape the legal
/// documents use — because a `ModalBottomSheet` is dismissible by gesture by construction
/// and refusing that gesture fights the component rather than using it.
@Composable
fun FrezIntroSheet(onNext: () -> Unit) {
    val palette = LocalGripPalette.current

    // Verbatim, in every locale: the catalog carries these four paragraphs with their
    // English as the French, which is the maker's ask and not an untranslated string.
    val paragraphs = listOf(
        tr("Hello, I’m Donghyun Kim, founder of Frez."),
        tr(
            "We keep our hardware margins low because we believe everyone should have " +
                "access to their own data. Frez Pro helps us develop new features and " +
                "provide reliable devices. In fact, Frez Dyno was made possible by our " +
                "early Pro subscribers.",
        ),
        tr(
            "If you enjoy using Frez Dyno and would like to support the future of Frez, " +
                "please consider trying Frez Pro.",
        ),
        tr("I hope Frez will be with you for many years to come."),
    )

    Dialog(
        // Nothing to answer: there is no dismissal this can be called for.
        onDismissRequest = {},
        properties = DialogProperties(
            usePlatformDefaultWidth = false,
            dismissOnBackPress = false,
            dismissOnClickOutside = false,
        ),
    ) {
        InstrumentSurface(
            modifier = Modifier.fillMaxSize(),
            shape = androidx.compose.ui.graphics.RectangleShape,
            color = palette.field,
            shadowElevation = 0.dp,
        ) {
            Column(
                Modifier
                    .safeDrawingPadding()
                    .fillMaxSize()
                    .readablePageWidth()
                    .padding(horizontal = Metrics.hPadding),
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                Text(
                    tr("A note from Frez"),
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.SemiBold,
                    color = palette.inkPrimary,
                    modifier = Modifier
                        .widthIn(max = Metrics.maxContentWidth)
                        .fillMaxWidth()
                        .padding(vertical = 14.dp),
                )
                Column(
                    Modifier
                        .weight(1f)
                        .verticalScroll(rememberScrollState()),
                ) {
                    InstrumentSurface(
                        modifier = Modifier
                            .widthIn(max = Metrics.maxContentWidth)
                            .fillMaxWidth(),
                    ) {
                        Column(
                            Modifier.padding(16.dp),
                            verticalArrangement = Arrangement.spacedBy(14.dp),
                        ) {
                            for (paragraph in paragraphs) {
                                Text(
                                    paragraph,
                                    style = MaterialTheme.typography.bodyMedium,
                                    color = palette.inkPrimary,
                                )
                            }
                            Column(verticalArrangement = Arrangement.spacedBy(2.dp)) {
                                Text(
                                    tr("Thank you,"),
                                    style = MaterialTheme.typography.bodyMedium,
                                    color = palette.inkPrimary,
                                )
                                Text(
                                    tr("Donghyun Kim"),
                                    style = MaterialTheme.typography.bodyMedium,
                                    color = palette.inkPrimary,
                                )
                            }
                        }
                    }
                }
                PrimaryButton(
                    tr("Next"),
                    modifier = Modifier
                        .widthIn(max = Metrics.maxContentWidth)
                        .fillMaxWidth()
                        .padding(vertical = 12.dp)
                        .testTag("frez.intro.next"),
                    onClick = onNext,
                )
            }
        }
    }
}
